File: python/logic/scoring_engine.py
import json
import numpy as np

class SalesScoringEngine:
    def __init__(self, rules_json_path):
        with open(rules_json_path, 'r', encoding='utf-8') as f:
            self.rules = json.load(f)
        
        self.weights = self.rules.get('weights_by_code', {})
        self.combos = self.rules.get('combo_rules', [])
        self.dimensions_list = list(self.rules.get('dimensions', {}).keys())
        self.actions_map = self.rules.get('default_actions_questions', {})

    def _get_intensity_multiplier(self, value, is_au=True):
        """
        Implementa s(I) = r(I)/5 conforme a equação.
        Se for AU: mapeia o valor contínuo para os degraus A-E.
        Se for Meta (Gaze/Head): s(vazio) = 1.0.
        """
        if not is_au:
            return 1.0 # s(vazio) = 1.0 para GAZE e HEAD
        
        # Mapeamento s(I) para intensidades FACS (A=0.2 a E=1.0)
        if value >= 0.8: return 1.0 # E
        if value >= 0.6: return 0.8 # D
        if value >= 0.4: return 0.6 # C
        if value >= 0.2: return 0.4 # B
        if value >= 0.1: return 0.2 # A
        return 0.0

    def process(self, input_payload):
        aus = input_payload.get("aus", {})
        meta = input_payload.get("meta", {})
        
        scores = {dim: 0.0 for dim in self.dimensions_list}
        active_codes = []

        # 1. Processar AUs: Σ s(Ii) * Wc,d
        for code, raw_val in aus.items():
            s_i = self._get_intensity_multiplier(raw_val, is_au=True)
            if s_i > 0 and code in self.weights:
                active_codes.append(code)
                for dim, weight in self.weights[code].items():
                    scores[dim] += s_i * weight

        # 2. Processar Meta (Gaze/Head): s(vazio)=1.0 * Wc,d
        # Tradução manual para os códigos do teu JSON
        meta_signals = []
        gaze = meta.get('gaze', 'CENTER')
        if gaze == "LOOKING_DOWN": meta_signals.append("GAZE64")
        
        pitch = meta.get('head_pitch', 0)
        if pitch > 15: meta_signals.append("HEAD54")
        elif pitch < -15: meta_signals.append("HEAD53")

        for m_code in meta_signals:
            if m_code in self.weights:
                active_codes.append(m_code)
                for dim, weight in self.weights[m_code].items():
                    scores[dim] += 1.0 * weight # s(vazio) = 1.0

        # 3. Aplicar Combos: Σ ak,d
        triggered_combos = []
        for combo in self.combos:
            if all(req in active_codes for req in combo['requires']):
                triggered_combos.append(combo['tag'])
                for dim, adj in combo.get('adjustments', {}).items():
                    scores[dim] += adj

        # 4. Clamp Final: clamp[-100, 100]
        final_scores = {k: max(-100, min(100, v)) for k, v in scores.items()}
        
        # Dimensão dominante para o HUD
        dom_dim = max(final_scores, key=lambda k: abs(final_scores[k]))
        
        return {
            "dominant_dimension": dom_dim,
            "dominant_value": int(final_scores[dom_dim]),
            "active_combos": triggered_combos,
            "scores": final_scores,
            "recommended_actions": [self.actions_map.get(dom_dim, {}).get('action', "Analise")],
            "questions": self.actions_map.get(dom_dim, {}).get('questions', [])
        }

File: typescript/logic/scoring_engine.ts
/**
 * Sales Scoring Engine
 * Processa regras do JSON de configuração e calcula pontuações para dimensões
 */

import type { ExpressionResult, FACSConfig, WindowPayload } from "@type/index";

export interface SalesScoringEngine {
	process(inputPayload: WindowPayload): ExpressionResult;
}

function getIntensityMultiplier(value: number, isAu: boolean = true): number {
	if (!isAu) {
		return 1.0;
	}

	if (value >= 0.8) return 1.0; // E
	if (value >= 0.6) return 0.8; // D
	if (value >= 0.4) return 0.6; // C
	if (value >= 0.2) return 0.4; // B
	if (value >= 0.1) return 0.2; // A
	return 0.0;
}

export function createSalesScoringEngine(
	rules: FACSConfig,
): SalesScoringEngine {
	const weights = rules.weights_by_code || {};
	const combos = rules.combo_rules || [];
	const dimensionsList = Object.keys(rules.dimensions || {});
	const actionsMap = rules.default_actions_questions || {};

	function process(inputPayload: WindowPayload): ExpressionResult {
		const aus = inputPayload.aus || {};
		const meta = inputPayload.meta || {};

		const scores: Record<string, number> = {};
		for (const dim of dimensionsList) {
			scores[dim] = 0.0;
		}

		const activeCodes: string[] = [];

		for (const [code, rawVal] of Object.entries(aus)) {
			const sI = getIntensityMultiplier(rawVal, true);
			if (sI > 0 && code in weights) {
				activeCodes.push(code);
				const weightMap = weights[code] || {};
				for (const [dim, weight] of Object.entries(weightMap)) {
					scores[dim] = (scores[dim] || 0) + sI * weight;
				}
			}
		}

		const metaSignals: string[] = [];
		const gaze = meta.gaze || "CENTER";
		if (gaze === "LOOKING_DOWN") metaSignals.push("GAZE64");

		const pitch = meta.head_pitch || 0;
		if (pitch > 15) metaSignals.push("HEAD54");
		else if (pitch < -15) metaSignals.push("HEAD53");

		for (const mCode of metaSignals) {
			if (mCode in weights) {
				activeCodes.push(mCode);
				const weightMap = weights[mCode] || {};
				for (const [dim, weight] of Object.entries(weightMap)) {
					scores[dim] = (scores[dim] || 0) + 1.0 * weight; // s(empty) = 1.0
				}
			}
		}

		const triggeredCombos: string[] = [];
		for (const combo of combos) {
			const requiresMet = combo.requires.every((req) =>
				activeCodes.includes(req),
			);
			const forbidsMet =
				combo.forbids && combo.forbids.length > 0
					? combo.forbids.some((forbid) => activeCodes.includes(forbid))
					: false;

			if (requiresMet && !forbidsMet) {
				triggeredCombos.push(combo.tag?.[0] || combo.id || "");
				for (const [dim, adj] of Object.entries(combo.adjustments || {})) {
					scores[dim] = (scores[dim] || 0) + adj;
				}
			}
		}

		const finalScores: Record<string, number> = {};
		for (const [k, v] of Object.entries(scores)) {
			finalScores[k] = Math.max(-100, Math.min(100, v));
		}

		let domDim = dimensionsList[0] || "engagement";
		let maxAbs = Math.abs(finalScores[domDim] || 0);
		for (const [k, v] of Object.entries(finalScores)) {
			const abs = Math.abs(v);
			if (abs > maxAbs) {
				maxAbs = abs;
				domDim = k;
			}
		}

		const actionData = actionsMap[domDim] || {
			action: "Analise",
			questions: [],
		};

		return {
			dominant_dimension: domDim,
			dominant_value: Math.round(finalScores[domDim] || 0),
			active_combos: triggeredCombos,
			scores: finalScores,
			recommended_actions: [actionData.action],
			questions: actionData.questions,
		};
	}

	return {
		process,
	};
}
